be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/ElectionValidator.scala

```diff
@@ -470,4 +534,126 @@ sealed class ElectionValidator(dbActorRef: => AskableActorRef) extends MessageDa
+      case reply                                   => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"ElectionValidator failed : unexpected DbActor reply '$reply'", rpcMessage.getId))
+    }
+  }
+
+  /** @param electionQuestionResult
+    *   The ElectionQuestionResult the ElectionValidator received.
+    * @return
+    *   The total number number of votes for the question.
+    */
+  private def countNumberOfVotes(electionQuestionResult: ElectionQuestionResult): Int = {
+    electionQuestionResult.result.map(_.count).sum
+  }
+
+  /** @param electionQuestionResult
+    *   The ElectionQuestionResult the ElectionValidator received.
+    * @return
+    *   True if all the ballot counts are greater that 0, false otherwise.
+    */
+  private def allBallotCountsArePositive(electionQuestionResult: ElectionQuestionResult): Boolean = {
+    electionQuestionResult.result.forall(electionBallotVotes => electionBallotVotes.count >= 0)
+  }
+
+  /** checks that the question ids received with the ResultElection are coherent with the ones received in the SetupElection
+    * @param rpcMessage
+    *   the rpc message to validate.
+    * @param questions
+    *   The list of ElectionQuestions that have been sent with the SetupElection.
+    * @param result
+    *   The list of ElectionQuestionResults the ElectionValidator received.
+    * @param error
+    *   The error to forward when the ballot options are incoherent.
+    * @return
+    *   GraphMessage: passes the rpcMessages to Right if successful else Left with pipeline error.
+    */
+  private def checkResultQuestionIds(
+      rpcMessage: JsonRpcRequest,
+      questions: List[ElectionQuestion],
+      result: List[ElectionQuestionResult],
+      error: PipelineError
+  ): GraphMessage = {
+    val setupElectionIds = questions.map(_.id).toSet
+    val resultElectionIds = result.map(_.id).toSet
+    if (setupElectionIds == resultElectionIds) {
+      Right(rpcMessage)
+    } else
+      Left(error)
+  }
+
+  /** checks that for the given ElectionResult, the ballot options match the ballot options sent in the SetupElection message.
+    * @param rpcMessage
+    *   The rpc message to validate.
+    * @param questions
+    *   The list of ElectionQuestions that have been sent with the SetupElection.
+    * @param result
+    *   The list of ElectionQuestionResults the ElectionValidator received.
+    * @param error
+    *   The error to forward in case where ballot options are incoherent.
+    * @return
+    *   GraphMessage: passes the rpcMessages to Right if successful else Left with pipeline error.
+    */
+  private def checkResultBallotOptions(
+      rpcMessage: JsonRpcRequest,
+      questions: List[ElectionQuestion],
+      result: List[ElectionQuestionResult],
+      error: PipelineError
+  ): GraphMessage = {
+    var isResultBallotValid = true
+    isResultBallotValid = result.forall(electionQuestionResult => {
```

Comment:
It is a nitpick, but you could use simply `question` instead of `electionQuestionResult`. It makes the code heavy imo.